core/resources.rs

```diff
@@ -42,11 +43,17 @@ impl ResourceTable {
     None
   }
 
-  // TODO: resource id allocation should probably be randomized for security.
+  // resource id allocation are randomized for security.
   fn next_rid(&mut self) -> ResourceId {
-    let next_rid = self.next_id;
-    self.next_id += 1;
-    next_rid as ResourceId
+    let mut rng = rand::thread_rng();
+    // rid's are cast into i32 then interpreted as unsigned number on typescript side.
+    // then when sending rid's back to Rust the cast back to i32 fails as it overflows.
+    // here we make the generated rid a valid positive i32 to prevent overflow
+    let mut next_rid = rng.gen::<u32>() & 0x7FFF_FFFF;
+    while self.map.contains_key(&next_rid) {
```

Comment:
I think it should panic if there is a collision.

Reply:
why? this would make the whole system unstable.